### PR TITLE
Prevent fields from being deleted from BaseModel.__dict__ when a field validator raises an unexpected exception

### PR DESCRIPTION
**Pull request by johnsabath**
## Change Summary

Addressing https://github.com/samuelcolvin/pydantic/issues/2044

## Related issue number

fix https://github.com/samuelcolvin/pydantic/issues/2044

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by johnsabath** [comment]
I could add some additional unit tests for this edge case as well, just let me know how you'd like to handle it.

File: pydantic/main.py [hunk 1]
```diff
@@ -384,7 +384,7 @@ def __setattr__(self, name, value):  # noqa: C901 (ignore complexity)
 
             known_field = self.__fields__.get(name, None)
             if known_field:
-                original_value = self.__dict__.pop(name)
+                original_value = self.__dict__[name]
```
Comment (PrettyWood) [inline pydantic/main.py:387]: We used to check `self.dict(exclude={name})` [before](https://github.com/samuelcolvin/pydantic/pull/2000/files#diff-8708af8d26c26644976866f974e0b810b0940ede43ff487e457057d5a1f07999L370) and I'm the one who suggested to just pop the value and use `self.__dict__` to be faster. I think we may have some undesired behaviour if we keep the whole dictionary since `values` will have the previous value and `v` will have the new one.
I'm trying to think on possible edge cases with validators that would use `values`.
If there are such cases we could use `self.dict(exclude={name})` or use a `try...except...`

**Comment by samuelcolvin** [comment]
what would be wrong with

```py
                try:
                    value, error_ = known_field.validate(value, self.__dict__, loc=name, cls=self.__class__)
                    if error_:
                        raise ValidationError([error_], self.__class__)
                except Exception:
                    self.__dict__[name] = original_value
                    raise
                else:
                    new_values[name] = value
```
?

File: pydantic/main.py (hunk 1 above)
Comment (johnsabath) [reply pydantic/main.py:387]: Ah, I see, that makes sense.

**Comment by johnsabath** [comment]
With @PrettyWood's additional context on why the field should still be removed from __dict__ before calling the validators, that seems like a good compromise to me @samuelcolvin

File: pydantic/main.py [hunk 2]
```diff
@@ -385,11 +385,15 @@ def __setattr__(self, name, value):  # noqa: C901 (ignore complexity)
             known_field = self.__fields__.get(name, None)
             if known_field:
                 original_value = self.__dict__.pop(name)
-                value, error_ = known_field.validate(value, self.__dict__, loc=name, cls=self.__class__)
-                if error_:
+                try:
+                    value, error_ = known_field.validate(value, self.__dict__, loc=name, cls=self.__class__)
+                    if error_:
+                        raise ValidationError([error_], self.__class__)
+                except Exception:
```
Comment (PrettyWood) [inline pydantic/main.py:392]: You should add a test for this

Comment (johnsabath) [reply pydantic/main.py:392]: on it

File: tests/test_validators.py [hunk 3]
```diff
@@ -1155,3 +1155,38 @@ class Model(BaseModel):
...
+        foo: str
+        bar: str
+
+        @validator('foo')
+        def validate_foo(cls, v, values):
+            nonlocal validators_values
+            validators_values = values
+
+    Model(foo='foo_value', bar='bar_value')
+    assert validators_values == {'bar': 'bar_value'}
```
Comment (PrettyWood) [inline tests/test_validators.py:1173]: There are two issues here:
- you don't test with `validate_assignement = True`
- you use the reference of `values` so it won't work as `values` is then updated before being checked with `assert`

Btw you could simply use a mock to check the validation. Something like
```python
def test_field_that_is_being_validated_is_excluded_from_validator_values(mocker):
    check_values = mocker.MagicMock()

    class Model(BaseModel):
        foo: str
        bar: str

        class Config:
            validate_assignment = True

        @validator('bar')
        def validate_bar(cls, v, values):
            check_values({**values})
            return v

    model = Model(foo='foo_value', bar='bar_value')
    check_values.reset_mock()

    model.bar = 'new_bar_value'
    check_values.assert_called_once_with({'foo': 'foo_value'})
```

Comment (johnsabath) [reply tests/test_validators.py]: Apologies, I didn't have my local venv set up, so these changes were untested.

Just pushed up a new set of changes now, and I used a mock to capture the calls like you suggested.

Comment (PrettyWood) [reply tests/test_validators.py]: No worries. Thanks a lot for this!

**Review by PrettyWood — APPROVED** [review]
LGTM!

File: tests/test_validators.py [hunk 4]
```diff
@@ -1155,3 +1155,43 @@ class Model(BaseModel):
...
+
+        @validator('foo')
+        def validate_foo(cls, v):
+            if v == 'raise_exception':
+                raise Exception()
+            return v
+
+    model = Model(foo='foo')
+    with pytest.raises(Exception):
+        model.foo = 'raise_exception'
```
Comment (samuelcolvin) [inline tests/test_validators.py:1196]:
```suggestion
                raise RuntimeError('test error')
            return v

    model = Model(foo='foo')
    with pytest.raises(RuntimeError, match='test error'):
        model.foo = 'raise_exception'
```

So we're sure we're catching the same error.

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM
